src/cleo/commands/completions_command.py

```diff
@@ -137,10 +138,32 @@ def render(self, shell: str) -> str:
 
         raise RuntimeError(f"Unrecognized shell: {shell}")
 
+    @staticmethod
+    def _get_prog_name_from_stack() -> str:
+        package_name = ""
+        frame = inspect.currentframe()
+        f_back = frame.f_back if frame is not None else None
+        f_globals = f_back.f_globals if f_back is not None else None
+        # break reference cycle
+        # https://docs.python.org/3/library/inspect.html#the-interpreter-stack
+        del frame
+
+        if f_globals is not None:
+            package_name = f_globals.get("__name__")
+
+            if package_name == "__main__":
+                package_name = f_globals.get("__package__")
+
+            if package_name:
+                package_name = package_name.partition(".")[0]
+
+        if not package_name:
+            raise CleoRuntimeError("Can not determine package name")
+
+        return package_name
+
     def _get_script_name_and_path(self) -> tuple[str, str]:
-        # FIXME: when generating completions via `python -m script completions`,
-        # we incorrectly infer `script_name` as `__main__.py`
-        script_name = self._io.input.script_name or inspect.stack()[-1][1]
+        script_name = self._io.input.script_name or self._get_prog_name_from_stack()
```

Comment:
did this work for you? ik it works for click, but would need tweaking

what about unifying the logic from the `help` command, which prefers `self._application.name` if defined? that feels more consistent and what I would expect: https://github.com/python-poetry/cleo/blob/7448f55fe931871488f20cbbdf68afbf12ce141d/src/cleo/commands/base_command.py#L61-L64

Reply:
I have only tested the `_get_prog_name_from_stack` method and it has returned the proper module name. `Aplication.name` is not always good, because you might have your program aliased on shell level (like having `pipx` installation with suffix).